kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/jvmbind.kt

```diff
@@ -2,86 +2,110 @@ package io.kotest.property.arbitrary
- * This can be used to control parts of the data generation, as well as including non-data classes in the
- * hierarchies for which an [Arb] is being derived.
+ * Callers should use [Arb.Companion.bind] without [KClass] and [KType] parameters instead.
+ */
+fun <T : Any> Arb.Companion.bind(providedArbs: Map<KClass<*>, Arb<*>>, kclass: KClass<T>, type: KType): Arb<T> {
+   val arb = Arb.forType(providedArbs, type)
+      ?: error("Could not locate generator for ${kclass.simpleName}, consider making it a dataclass or provide an Arb for it.")
+   return arb as Arb<T>
+}
+
+/**
+ * **Do not call directly**
  *
- * Note: This method only supports parameters the following kind of parameter:
- * - Data classes, where all properties also fall into this category
- * - Pair, where 1st and 2nd fall into this category
- * - Primitives
- * - Enums
- * - LocalDate, LocalDateTime, LocalTime, Period, Instant
- * - BigDecimal, BigInteger
- * - Collections (Set, List, Map)
+ * Callers should use [Arb.Companion.bind] without [KClass] and [KType] parameters instead.
  */
+@Deprecated(
+   "Superceded by bind without KClass parameter.",
+   ReplaceWith("bind(providedArbs)")
+)
 fun <T : Any> Arb.Companion.bind(providedArbs: Map<KClass<*>, Arb<*>>, kclass: KClass<T>): Arb<T> {
+   return forClassUsingConstructor(providedArbs, kclass)
+}
+
+internal fun <T : Any> Arb.Companion.forClassUsingConstructor(
+   providedArbs: Map<KClass<*>, Arb<*>>,
+   kclass: KClass<T>
+): Arb<T> {
+   val className = kclass.qualifiedName ?: kclass.simpleName
    val constructor = kclass.primaryConstructor ?: error("could not locate a primary constructor")
-   check(constructor.parameters.isNotEmpty()) { "${kclass.qualifiedName} constructor must contain at least 1 parameter" }
+   check(kclass.visibility != KVisibility.PRIVATE) { "The class $className must be public." }
+   check(constructor.visibility != KVisibility.PRIVATE) { "The primary constructor of $className must be public." }
+
+   if (constructor.parameters.isEmpty()) {
+      return Arb.constant(constructor.call())
+   }
 
    val arbs: List<Arb<*>> = constructor.parameters.map { param ->
-      val arb = Arb.forType(providedArbs, param.type)
-         ?: error("Could not locate generator for parameter ${kclass.qualifiedName}.${param.name}")
+      val arb = arbForParameter(providedArbs, className, param)
       if (param.type.isMarkedNullable) arb.orNull() else arb
    }
 
    return Arb.bind(arbs) { params -> constructor.call(*params.toTypedArray()) }
 }
 
+private fun arbForParameter(providedArbs: Map<KClass<*>, Arb<*>>, className: String?, param: KParameter): Arb<*> {
+   val errorMsg = "Could not locate generator for parameter $className.${param.name}, consider providing an Arb for it."
+   val arb =
+      try {
+         Arb.forType(providedArbs, param.type)
+      } catch (e: IllegalStateException) {
+         throw IllegalStateException(errorMsg, e)
+      }
+   return arb ?: error(errorMsg)
+}
```

Comment:
Could be reduced to this I believe? Unless I miss something in `Arb.forType` that might throw `IllegalStateException`?

```suggestion
private fun arbForParameter(providedArbs: Map<KClass<*>, Arb<*>>, className: String?, param: KParameter): Arb<*> =
   checkNotNull(Arb.forType(providedArbs, param.type)) { "Could not locate generator for parameter $className.${param.name}, consider providing an Arb for it." }
```

Added benefit: Error message is created lazily 🙂

Reply:
For parameters that are themselves classes `targetDefaultForType` will call `forClassUsingConstructor` recursively. An issue with the primary constructor then throws IllegalsStateException. But since the parameter name is lost in that context I figured it worthwhile to rethrow the exception.

See `ArbDefaultsTest` "throw on complex data class".

I agree this is not pretty though. I would like to see this improved too.

Reply:
I've added a test for a nested parameter exception, and changed the exception message what. 
I hope it makes more sense now.

Reply:
Thanks for clarifying!